src/libgit2/grafts.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "grafts.h"
+
+#include "futils.h"
+#include "oidarray.h"
+#include "parse.h"
+
+bool git_shallow__enabled = false;
+
+struct git_grafts {
+	/* Map of `git_commit_graft`s */
+	git_oidmap *commits;
+
+	/* File backing the graft. NULL if it's an in-memory graft */
+	char *path;
+	git_oid path_checksum;
+};
+
+int git_grafts_new(git_grafts **out)
+{
+	git_grafts *grafts;
+
+	grafts = git__calloc(1, sizeof(*grafts));
+	GIT_ERROR_CHECK_ALLOC(grafts);
+
+	if ((git_oidmap_new(&grafts->commits)) < 0) {
+		git__free(grafts);
+		return -1;
+	}
+
+	*out = grafts;
+	return 0;
+}
+
+int git_grafts_from_file(git_grafts **out, const char *path)
+{
+	git_grafts *grafts = NULL;
+	int error;
+
+	if ((error = git_grafts_new(&grafts)) < 0)
+		goto error;
+
+	grafts->path = git__strdup(path);
+	GIT_ERROR_CHECK_ALLOC(grafts->path);
+
+	if ((error = git_grafts_refresh(grafts)) < 0)
+		goto error;
+
+	*out = grafts;
+error:
+	if (error < 0)
+		git_grafts_free(grafts);
+	return error;
+}
+
+void git_grafts_free(git_grafts *grafts)
+{
+	if (!grafts)
+		return;
+	git__free(grafts->path);
+	git_grafts_clear(grafts);
+	git_oidmap_free(grafts->commits);
+	git__free(grafts);
+}
+
+void git_grafts_clear(git_grafts *grafts)
+{
+	git_commit_graft *graft;
+
+	assert(grafts);
+
+	git_oidmap_foreach_value(grafts->commits, graft, {
+		git__free(graft->parents.ptr);
+		git__free(graft);
+	});
+
+	git_oidmap_clear(grafts->commits);
+}
+
+int git_grafts_refresh(git_grafts *grafts)
+{
+	git_str contents = GIT_STR_INIT;
+	int error, updated = 0;
+
+	assert(grafts);
+
+	if (!grafts->path)
+		return 0;
+
+	error = git_futils_readbuffer_updated(&contents, grafts->path,
```

Comment:
This could be clearer:

```c
	if (error = git_futils_readbuffer_updated(&contents, grafts->path,
					      (grafts->path_checksum).id, &updated) < 0) {
		if (error == GIT_ENOTFOUND) {
			git_grafts_clear(grafts);
			error = 0;
		}
		goto cleanup;
	}

	if (!updated) {
		goto cleanup;
	}
```